API Support

gokmen opened the pull request:
## Description
This PR will provide REST api for Koding Node backend with proper cookie or token or for public requests. **This is currently experimental and wip.**

## Screenshots (if appropriate):

![](http://g.recordit.co/AvGdOFRfu6.gif)

Full screencast: http://recordit.co/AvGdOFRfu6

## TODO
 - [x] Add support for descriptions in schema on `Bongo::Model`
 - [x] Generate initial model definitions for Swagger api generator from Bongo models
 - [x] Generate api endpoints for static methods of models
 - [x] Generate api endpoints for instance methods of models
 - [x] Remote API handler for initial requests
 - [x] Enable throttling on remote endpoint
 - [x] Setup Swagger-UI in place to provide auto generated api docs
 - [x] Auto generate swagger.yaml on model change, serve it under website
 - [x] Provide a way to login on apidocs ui and embed `clientId` into it
 - [x] Support static methods of shared models
 - [x] Support instance methods of shared models
 - [x] Use existing `clientId` from cookies as authorization token
 - [x] Fix unhandled Bongo calls 
 - [x] Fail when invalid signatures provided
 - [x] Provide possible valid signatures with validation error
 - [x] Provide search functionality in api docs viewer
 - [x] Generate parameter types, examples and description from docs
 - [x] Compare source code and swagger.json against changes
  _example failed build https://app.wercker.com/buildstep/58346a4b62c4510100c29b8a if it has outdated swagger.json_
  - [ ] Tests needs to be written for the api cc/ @usirin

It has now embed documents from source code feature in this PR uses Codo library https://github.com/coffeedoc/codo we will eventually use this to generate human readable code documentation as well.

![image](https://cloud.githubusercontent.com/assets/42368/20435385/928be4d0-adb4-11e6-9b8a-976c5c13283b.png)

gokmen committed:
Bongo: Model: add _description support for fields

gokmen committed:
Models: JStackTemplate: add _descriptions per field, as example.

gokmen committed:
Bongo: Model: describeSchema to provide more information per field

Before it was only providing type of the field, now it produces all the
information including required, description and itemtype for array types

gokmen committed:
Server: Client: remove not used sections

gokmen committed:
Models: update Array field types to a proper one with itemType including

gokmen committed:
Remote: api handler for rest api, all in wip - this is initial commit

gokmen committed:
Scripts/Api-Generator: initial commit for swagger yaml generator

usirin commented:
![](https://media.giphy.com/media/amkxRknWsbWtG/giphy.gif)

gokmen committed:
JAccount: reserveNames: remove outdated reserve helper

gokmen committed:
website/apidocs: Add Swagger-UI for apidocs

gokmen committed:
Apidocs: styling update for api docs

gokmen committed:
Apidocs: update SwaggerUI instance for api doc needs

gokmen committed:
Expressify: check on callbacks, provide option to processPayload before send

gokmen committed:
Server: remove /-/api/remote endpoint from web server worker

gokmen committed:
RemoteAPI Initial version under /remote.api/ in socialworker

gokmen committed:
JAccount: remove deprecated sharedMethods

gokmen committed:
Scripts/Api-Generator: update with detailed api endpoints examples

gokmen committed:
RemoteAPI: fix cyclomatic complexity by splitting down into small pieces

gokmen committed:
RemoteAPI: use body as is if it's provided as an Array

gokmen committed:
cleanup: remove go/bin as well

gokmen committed:
Bongo: invokeMethod: fix all the unhandled calls issues

This is very annoying that I just realised all the calls drops here if
there is an error in the invoker. For example, if you make an instance
call on a non-existent instance it never calls the callback. Or if you
try to call a non-existent method over any Bongo.model this never calls
the callback as well. It was valid for lots of other cases as well in
this commit.

With this commit we will call the provided callback with a proper error
instead. which I assume will fix the remaining memory issues on Bongo.

gokmen committed:
Bongo: Base: testSignature to return possible signatures with the result

gokmen committed:
Social: add a generic error handler for unhandled exceptions

gokmen committed:
RemoteApi: no need to pass group info, it will be taken from sessionToken

gokmen committed:
RemoteApi: provide possible signatures if a call is not valid

gokmen committed:
JComputeStack: create: check session and group existence on the client

gokmen committed:
ApiDocs: add Cookies helper

gokmen committed:
ApiDocs: to use existing clientId for Authorization on Swagger

gokmen committed:
RemoteApi: better suggestions for calls with unrecognised signatures

gokmen committed:
JAccount: byRelevance: remove deprecated signature

gokmen committed:
JSession: activeSession: to fetch active session data

gokmen committed:
SetCSRFToken: shouldn't continue to next

gokmen committed:
ApiGenerator: update for default body parameter

gokmen committed:
ApiDocs: Login/Logout functionality with extensive API usage with auth data

gokmen committed:
RemoteApi: check all types of methods signatures not only statics

gokmen committed:
ApiDocs: SwaggerUI cleanup, add Koding logo, use Roboto fonts

gokmen committed:
Bongo: Base: use statik, check against signature existence on testSignature

gokmen committed:
JApiToken: remove deprecated signature of ::create

gokmen committed:
SocialWorker: handle all errors under social worker

gokmen committed:
ApiGenerator: fix linter issues, update styling on apidocs

gokmen committed:
ApiGenerator:add definitions for default request and responses

gokmen committed:
ApiGenerator: generate API endpoints for both static and instance methods

gokmen committed:
RemoteApi: verify session and use group information from session

gokmen committed:
Swagger schema: Add auto-generated swagger definition for feature usages

gokmen committed:
ApiDocs: keep header fixed on top, remove auth method on logout

gokmen committed:
ApiGenerator: remove predefined paths, update swagger.json

gokmen committed:
ApiGenerator: add unauthorizedRequest type make required optional

cihangir commented on node_modules_koding/bongo/lib/bongo.coffee:
```diff
@@ -495,62 +499,84 @@ module.exports = class Bongo extends EventEmitter
+
+  handleUserError: (message, args) ->
+
+    console.warn '[Bongo:UserError]', message, args
```
👍

cihangir commented on node_modules_koding/bongo/lib/expressify.coffee:
```diff
@@ -134,13 +138,20 @@ module.exports = (options = {}) ->
-            if _req.method is 'authenticateUser'
+            unless _req.callbacks
+              healthyQueue = no
+              return res.status(500).send 'An error occcurred'
```
what about a more informative message or a log statement?

cihangir commented on workers/social/lib/social/models/account.coffee:
```diff
@@ -56,28 +56,14 @@ module.exports = class JAccount extends jraphical.Module
         ]
         some:
           (signature Object, Object, Function)
-        cursor:
```
👍

cihangir commented:
could you please make sure swagger.json is always up-to-date with the current codebase in CI env?

szkl commented on workers/social/lib/social/models/session.coffee:
```diff
@@ -57,6 +61,10 @@ module.exports = class JSession extends Model
 
 
+  @activeSession = secure (client, callback) ->
+    @one { clientId: client.sessionToken }, callback
```
Would it be good to match username and group too?

gokmen replied:
I don't have those information, just getting token from header. But using the `username` and `group` in returned `JSession` object.

szkl replied:
Okay. I thought `client` would have them.

gokmen replied:
ah sorry, I thought you were asking for some place else, you're right I can add that check here. 👊

gokmen commented:
> could you please make sure swagger.json is always up-to-date with the current codebase in CI env?

@cihangir sure thing! At first I was planning to not to keep swagger.json in repo but then I thought it might be a good thing to see what is changed on api. I'll add a wercker step to verify swagger.json with the code.

cihangir commented:
I also liked the idea of keeping swagger def in the repo 👍  💯

gokmen committed:
Models: updated comments for upcoming api description feature over comments

gokmen committed:
Expressify: if a broken request coming through here, clear clientId from cookie

gokmen committed:
Models: Add more sample documentation

gokmen committed:
Scripts: DocGen: it will parse documents from given folder/path

For example it will generate following data for ProviderInterface file:

https://github.com/koding/koding/blob/remote/workers/social/lib/social/models/computeproviders/providerinterface.coffee

λ ~/Code/koding/ remote* coffee scripts/docgen.coffee workers/social/lib/social/models/computeproviders/providerinterface.coffee
{
  "ProviderInterface": {
    "description": "Base class for all providers",
    "parameters": [],
    "examples": [
      {
        "title": "How to subclass a provider",
        "code": "class Aws extends ProviderInterface\n\n  @providerSlug  = 'aws'\n  @bootstrapKeys = ['key_pair', 'rtb', 'acl']\n  @sensitiveKeys = ['access_key', 'secret_key']"
      }
    ],
    "instance": {},
    "static": {
      "fetchCredentialData": {
        "description": "Generic fetcher for JCredential's data on this provider",
        "parameters": [
          {
            "type": "Object",
            "name": "client",
            "description": "valid client object"
          },
          {
            "type": "Object",
            "name": "credential",
            "description": "valid JCredential instance"
          },
          {
            "type": "Function",
            "name": "callback",
            "description": "function"
          }
        ],
        "returns": {
          "type": "Boolean, Object",
          "description": "credential data"
        },
        "options": {}
      },
      "update": {
        "description": "Generic modifier for JMachine's on this provider",
        "parameters": [
          {
            "type": "Object",
            "name": "valid",
            "description": "client object"
          },
          {
            "type": "Object",
            "name": "options",
            "description": "for update"
          }
        ],
        "returns": {
          "type": "Boolean, Object",
          "description": "modified JMachine instance"
        },
        "options": {
          "options": [
            {
              "type": "String",
              "name": "machineId",
              "description": "target JMachine._id"
            },
            {
              "type": "Boolean",
              "name": "alwaysOn",
              "description": "always on state of the machine"
            }
          ]
        }
      }
    }
  }
}

gokmen committed:
Bongo: keep track of registered model's files for latter use

gokmen committed:
ApiGenerator: default definitions, better naming

gokmen committed:
ApiGenerator: use docGen to get description of methods, classes

gokmen committed:
Models: update comments for some examples

gokmen committed:
Swagger.json: updated with latest changes from source files

gokmen committed:
ApiDocs: sync swagger-ui with latest koding/swagger-ui

gokmen committed:
ApiDocs: add search functionality

gokmen committed:
ComputeProvider: fix cyclomatic complexity lint issue

gokmen committed:
DocGen: generate json schema for parsed examples

gokmen committed:
ComputeProvider: add api examples for create and ping methods with docs

gokmen committed:
ApiGenerator: include examples and schema from docs, update swagger

gokmen committed:
Expressify: add some comments for mission critical places

gokmen committed:
JSession:activeSession: check username on client object with in selector

gokmen committed:
ApiDocs: sync with koding/swagger-ui

gokmen committed:
RemoteApi: enable rate limiting for api routes

gokmen committed:
ApiGenerator: fix linter issues

gokmen committed:
DocGen: use inline schema generator for custom needs, remove deps

gokmen committed:
ApiGenerator: add --check option to compare source code with swagger.json

gokmen committed:
Wercker: check api consistency and fail if it's outdated

gokmen committed:
Wercker: check api consistency on test instance

gokmen committed:
Api: Update swagger.json

gokmen commented:
I'm done with this shall we merge? cc/ @cihangir @usirin @szkl @rjeczalik

gokmen committed:
RemoteApi: return possible signatures as error.signatures

cihangir commented on workers/social/lib/social/remoteapi.coffee:
```diff
@@ -0,0 +1,188 @@
+KONFIG    = require 'koding-config-manager'
```
could you write test for this file?

cihangir commented on workers/social/lib/social/models/session.coffee:
```diff
@@ -57,6 +61,18 @@ module.exports = class JSession extends Model
+    if not username or not clientId
+      return callback new KodingError 'Access denied'
+
+    @one { username, clientId }, callback
```
clientId is enough, we dont need username + we dont have index on these two

cihangir commented on workers/social/lib/social/models/computeproviders/providerinterface.coffee:
```diff
@@ -11,8 +11,17 @@ NOT_IMPLEMENTED = ->
+
+# Base class for all providers
+#
+# @example How to subclass a provider
```
👍

cihangir commented on wercker.yml:
```diff
@@ -76,6 +76,9 @@ build:
           ./configure --config sandbox --projectRoot $WERCKER_SOURCE_DIR --ebEnvName $EB_ENV_NAME --version $(cat $WERCKER_ROOT/VERSION)
     - script:
+        name: check api consistency
+        code: scripts/wercker/run-command ./run exec scripts/api-generator.coffee --check
```
👍

gokmen committed:
JSession::activeSession: query by clientId only

cihangir reviewed (approved):
💯
